### ISSUE TITLE
[feature] Better Cache API

### ISSUE THREAD

**Issue by wsargent** [playframework/playframework#5912]
Add a better Cache API that returns Future[T] and isn't tied to EhCache.

* https://github.com/playframework/playframework/issues/4972
* https://github.com/playframework/playframework/issues/5296

**Comment by gmethvin** [comment]
There are actually three separate issues:
 - Make `CacheAPI` return `Future[T]` where possible.
 - Separate cache API module from implementation.
 - Use a different default cache implementation (perhaps [caffeine](https://github.com/ben-manes/caffeine)).

Also, we may want to consider creating modules for other cache implementations where possible. Scalacache has a few we don't have and the implementations are fairly straightforward.

**Comment by marcospereira** [comment]
Any reasons to not adopt ScalaCache? We can do something like we did to [support different database pools](https://github.com/marcospereira/playframework/blob/58a150d712bdf3bf4d14709569afcfd9ea97fb7f/framework/src/play-jdbc/src/main/scala/play/api/db/DBModule.scala#L65-L65) and have a provider to decide between cache implementations supported by ScalaCache.

**Comment by gmethvin** [comment]
@marcospereira See the discussion on #5296. I think we want to keep our current API (which is vastly different from ScalaCache's), and I feel like cache implementations are simple enough that we could create our own play modules for them.

**Comment by wsargent** [comment]
Also relevant: https://github.com/playframework/playframework/pull/5923

**Comment by gmethvin** [comment]
Other ideas:

- We should move the cache API into a separate play-cache repo, since not much of it needs to be tied to the Play release cycle. We can have several different implementations there as well, similar to how [scalacache](https://github.com/cb372/scalacache) does it.

- It should be easy to bind different cache implementations for different caches. We currently can use `@Named` annotations for different ehcache implementations but it should be simple to, for example, use caffeine for one cache and memcached for another. I'm also not sure if using an annotation is the best way; it would be nice if that information were encoded in the type instead, perhaps using a [Tagged type](http://eed3si9n.com/learning-scalaz/Tagged+type.html).

**Comment by charlesxucheng** [comment]
I hope Pay will have a higher level cache API similar to JSR 107 (saw your comments about lack of async support and bad review on JSR 107 - so could you come out with something similar?) or something like ScalaCache's memoize function feature.

Basically I found it a bit redundant to put all the cache put and get calls in the code, plus it clutters the code (cross-cutting concern). That's why I used ScalaCache instead of Play CacheAPI for now. However, ScalaCache has its own limitations (can't easily turn it off for testing, doesn't work well with Guice in my case)

**Comment by adamrabung** [comment]
By return Option, callers can't distinguish between cache miss & IO errors reaching the cache. This is problematic for some types of caches. It would be really ugly, but a Either[String, Option[T]] version of get would be really helpful.

**Comment by ben-manes** [comment]
Would there be interest in moving off Ehcache 2 as the default implementation? That version was obsoleted by v3, which probably wasn't adopted due to not supporting explicit expiration settings. Both are fairly slow and inefficient from my analysis.

I'm about to release Caffeine 2.6, which extends the API to capture your use-case of setting the expiration explicitly in a `put` call. The policy-specific extension point is `cache.policy().expireVariably()`, where this release adds `put(key, value, duration, unit)` and the `putIfAbsent` equivalent. This matches the needs in your current implementation.

I think this would be a cleaner upgrade path and the richer API should provide more flexibility as your APIs evolve.

**Comment by gmethvin** [comment]
@ben-manes I agree caffeine is a better option than ehcache. We've already separated the cache api from the implementation, so we need to create a `play-caffeine` module that implements the cache API. I created issue #7974 for this.

**Comment by bmontuelle** [comment]
Hi, I'm having a bit of struggle to clear cached entries that could use multiple parameters (complex SQL search query results). 
Key is composed of all the optional parameters, and a prefix for each one. So when I need to purge entries I need to generate all possible keys containing a value related to the database entries being updated/deleted.
I could have a great use of something like this
`CacheApi.removeEntriesContaining(keyContaining: String, otherKeysContaining: String*): Unit`
It would remove every entry whose key contains substrings passed as parameters

Or maybe even
`CacheApi.removeEntriesMatching(pattern: Pattern): Unit`
To increase flexibility in removing entries

I know something like this could represent a bit of work, as it is not already provided by Caffeine or EHCache, Its implementation might also present performance issue if we rely on the keys list to match pattern on every one of them.

**Comment by ben-manes** [comment]
@bmontuelle Caffeine provides an `asMap()` view for iterations, such as `cache.asMap().keySet().removeIf(predicate)` might work for you. I believe Ehcache2's `Cache.getKeysWithExpiryCheck()` could be used as an equivalent, but is a more expensive copy rather than a live view. Both would require an O(n) iterations to filter the keys by, so your other options are to keep an auxiliary data structure in sync that uses sorting for fancier queries. In Caffeine that could be done via map computations, whereas Ehcache2 might do so via loader/writer/listener hooks.
